drivers/volume/portworx/portworx.go

```diff
@@ -510,174 +514,1520 @@ func (d *portworx) RecoverDriver(n node.Node) error {
 			}
-		case api.SpecEphemeral:
-			if requestedSpec.Ephemeral != vol.Spec.Ephemeral {
-				return errFailedToInspectVolume(volumeName, k, requestedSpec.Ephemeral, vol.Spec.Ephemeral)
+		}
+		logrus.Infof("Successfully inspected volume: %v (%v)", vol.Locator.Name, vol.Id)
+
+	}
+
+	return nil
+}
+
+// Validate the volume replicas as per VolumePlacementStrategy rule applied to the volume
+func (d *portworx) ValidateVps(vol *api.Volume, appVols map[string]map[string]string, volNodes []api.StorageNode, nodeReplMap map[string][]*api.Volume, appVolNodes map[string]api.StorageNode) error {
+
+	logrus.Debugf("Volume details: %v ===\n (%v) Volumes per Node List:%v, all nodes of the App's Volumes: %v ", vol, appVols, nodeReplMap, appVolNodes)
+	logrus.Infof("Validate VPS  for Volume:%v ,  VPS Rule :(%v)", vol.Id, vol.Spec.GetPlacementStrategy())
+
+	if vpsrule := vol.Spec.GetPlacementStrategy(); vpsrule != nil {
+
+		// Get Volume Labels
+		volLabels := vol.Spec.GetVolumeLabels()
+
+		//Get vps spec from k8s
+		volVpsRule, err := k8sTalisman.GetVolumePlacementStrategy(volLabels["placement_strategy"])
+		logrus.Debugf("====Volume details-5: %v ===\n Placement strategy (%v)", vol.Id, volVpsRule)
+
+		if err != nil {
+			return err
+		}
+
+		// For each VPS  group rules
+		// Group1 -Replica Affinity Group
+		if volVpsRule.Spec.ReplicaAffinity != nil {
+			for _, rRule := range volVpsRule.Spec.ReplicaAffinity {
+
+				logrus.Infof("Validate Replica Affinity Rule Vol:%v === RA: Spec %v", vol.Id, rRule.CommonPlacementSpec)
+				err = d.ValidateReplicaAffinity(vol, rRule, volNodes)
+				if err != nil {
+					return &ErrFailedToInspectVolume{
+						ID:    vol.Id,
+						Cause: fmt.Sprintf("failed to validate ReplicaAffinity. Err: %v", err),
+					}
+				}
 			}
-		case api.SpecFilesystem:
-			if requestedSpec.Format != vol.Spec.Format {
-				return errFailedToInspectVolume(volumeName, k, requestedSpec.Format, vol.Spec.Format)
+		}
+		// Group2 -Replica Anti Affinity Group
+		if volVpsRule.Spec.ReplicaAntiAffinity != nil {
+			for _, raRule := range volVpsRule.Spec.ReplicaAntiAffinity {
+
+				logrus.Infof("Validate Replica Anti Affinity Rule Vol:%v === RAA: Spec %v", vol.Id, raRule.CommonPlacementSpec)
+				err = d.ValidateReplicaAntiAffinity(vol, raRule, volNodes)
+				if err != nil {
+					return &ErrFailedToInspectVolume{
+						ID:    vol.Id,
+						Cause: fmt.Sprintf("failed to validate ReplicaAntiAffinity. Err: %v", err),
+					}
+				}
 			}
-		case api.SpecBlockSize:
-			if requestedSpec.BlockSize != vol.Spec.BlockSize {
-				return errFailedToInspectVolume(volumeName, k, requestedSpec.BlockSize, vol.Spec.BlockSize)
+		}
+		// Group3 -Volume Affinity Group
+		if volVpsRule.Spec.VolumeAffinity != nil {
+			for _, vRule := range volVpsRule.Spec.VolumeAffinity {
+
+				logrus.Infof("Validate Volume Affinity Rule Vol:%v === VA: Spec %v", vol.Id, vRule)
+				err = d.ValidateVolumeAffinity(vol, vRule, volNodes, nodeReplMap, appVolNodes)
```

Comment:
I don't see the part where we are looking at multiple volumes together and then checking if they satisfied the affinity requirements. `ValidateVolumeAffinity` is only being called on a single volume.